src/server/transaction.cc

```diff
@@ -838,13 +838,23 @@ OpStatus Transaction::ScheduleSingleHop(RunnableType cb) {
 
 // Runs in coordinator thread.
 void Transaction::Execute(RunnableType cb, bool conclude) {
+  auto tracking_wrap = [cb, this](Transaction* t, EngineShard* shard) -> RunnableResult {
```

Comment:
@dranikpg this with the changes in `InvokeCmd` seemed to be the most non intrusive way (to comply with the requirements of the state machine)

Reply:
1. I would like to understand why tracking requires transaction semantics (an example will be fine)
2. why cid_ is not enough and we need `invoke_cid_` ?

Reply:
> I would like to understand why tracking requires transaction semantics (an example will be fine)

Because invalidation messages must be sent before the transaction concludes. Otherwise, we might accidentally skip them. An example would be:

```
>> CLIENT TRACKING ON
>> GET FOO
>> SET FOO BAR 
>> GET FOO
>> SET FOO BAR
>> GET FOO ---------> might miss Invalidation message
```

A valid execution would be once we call the first `SET` we will send an invalidation message as a separate transaction. Now before that even starts/concludes, the `GET` that follows will get executed first and it will itself issue a separate transaction to send an invalidation message. Now the problem here is, that once we send an invalidation message we remove the key from the tracking map (since we only send invalidation messages once until the key is reread). Then the second invalidation transaction won't work because the key no longer exists in the map and we will never get that second invalidation message.